cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitor.java

```diff
@@ -186,10 +230,24 @@ protected void run() {
             throw new ServiceUnavailableException(response.getResponseBodyAsString());
           }
 
-          Map<String, List<MonitorMessage>> monitorResponses =
+          Map<String, Deque<MonitorMessage>> monitorResponses =
             GSON.fromJson(response.getResponseBodyAsString(StandardCharsets.UTF_8), MAP_STRING_MESSAGE_TYPE);
 
-          long latestPublishTime = processResponse(monitorResponses);
+          // Update ProgramFinishTime when remote runtime is in terminal state. Also buffer all the program status
+          // events. This is done before transactional publishing to avoid refetching same remote runtime status
+          // messages.
+          updateProgramFinishTime(monitorResponses);
+
+          startTime = System.currentTimeMillis();
```

Comment:
You shouldn't update the start time here. You should update it after everything is finished successfully and after the sleep.